### PR TITLE
Prompt user with info about terminal config, add settings search by @id, and add sendToShell setting

### PR DESCRIPTION
**Pull request by meganrogge**
Use @id: to search for one or more comma separated settings ids

Use keyboard shortcuts on the terminal corresponding to `terminal.integrated.commandsToSkipShell` commands and see an alert to configure terminal settings. Clicking `Configure` or `Don't show again` should result in never seeing the prompt again as the same user. Clicking the x on the prompt will show the prompt again the next time it's triggered.

Set `terminal.integrated.sendToShell`: true and use any key-bindings that trigger something to happen in the terminal. They should work and not be handled by the workbench. For example, Ctrl+E should jump to the end of the line instead of opening the quick pick. 

You should still be able to use Ctrl+1-9 to navigate around the editor groups.

### CODE REVIEW

File: src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts [hunk 1]
```diff
@@ -312,6 +313,18 @@ export class SettingsTreeSettingElement extends SettingsTreeElement {
...
+	matchesAnyId(idFilters?: Set<string>): boolean {
+		if (!idFilters || !idFilters.size) {
+			return true;
+		}
+
+		if (!this.setting.key) {
+			return false;
+		}
+
+		return Array.from(idFilters).some(id => id === this.setting.key);
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts:326]: You should be able to just check the set directly, like `idFilters.has(...`

File: src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts [hunk 2]
```diff
@@ -15,6 +15,11 @@ export const terminalConfiguration: IConfigurationNode = {
 	title: localize('terminalIntegratedConfigurationTitle', "Integrated Terminal"),
 	type: 'object',
 	properties: {
+		'terminal.integrated.sendToShell': {
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts:18]: I wonder whether it should say something about keys, like `sendKeybindingsToShell`?

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 3]
```diff
@@ -541,10 +543,30 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
 			}
-
-			// Skip processing by xterm.js of keyboard events that resolve to commands described
-			// within commandsToSkipShell
-			if (resolveResult && this._skipTerminalCommands.some(k => k === resolveResult.commandId)) {
+			const promptChoices: IPromptChoice[] = [
+				{
+					label: nls.localize('configureTerminalSettings', "Configure Terminal Settings"),
+					run: () => {
+						this._commandService.executeCommand('workbench.action.openSettings', '@id:terminal.integrated.commandsToSkipShell,terminal.integrated.sendToShell,terminal.integrated.allowChords');
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:550]: I think there is a method on PreferencesService that will let you do this, you shouldn't have to call executeCommand most of the time

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 4]
```diff
@@ -541,10 +543,30 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 				event.preventDefault();
 				return false;
 			}
-
-			// Skip processing by xterm.js of keyboard events that resolve to commands described
-			// within commandsToSkipShell
-			if (resolveResult && this._skipTerminalCommands.some(k => k === resolveResult.commandId)) {
+			const promptChoices: IPromptChoice[] = [
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:546]: Move this inside the if block so we don't create these objects on every keypress unless needed

**Review by roblourens — COMMENTED** [review]
Built this and tried it out and I'm wondering, maybe I should only ever see this notification once. It's really easy to trigger, and my muscle memory is just to press escape when I want it to go away. Then I end up seeing it over and over.

**Comment by roblourens** [comment]
You don't have to keep merging master into your PR, unless there are conflicts to resolve or some new change that you want to base this on.

**Comment by meganrogge** [comment]
> Built this and tried it out and I'm wondering, maybe I should only ever see this notification once. It's really easy to trigger, and my muscle memory is just to press escape when I want it to go away. Then I end up seeing it over and over.

@roblourens yes right now it only closes if you select a prompt choice. I can make it never show again if you close the prompt if that's how most of them work.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 5]
```diff
@@ -547,7 +547,10 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 			// for keyboard events that resolve to commands described
 			// within commandsToSkipShell, either alert or skip processing by xterm.js
 			if (resolveResult && this._skipTerminalCommands.some(k => k === resolveResult.commandId) && !this._configHelper.config.sendKeybindingsToShell) {
-				if (this._storageService.getBoolean(SHOW_TERMINAL_CONFIG_PROMPT, StorageScope.GLOBAL, true)) {
+				if (this._storageService.getBoolean(SHOW_TERMINAL_CONFIG_PROMPT, StorageScope.GLOBAL, true) &&
+					resolveResult.commandId !== 'workbench.action.terminal.toggleTerminal' &&
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:551]: A possible issue with this. If I have focus in the terminal, and I press a keybinding bound to one of these, and I expect it to go to the terminal but it triggers the command. That's actually the exact scenario we are trying to capture.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 6]
```diff
@@ -542,9 +544,27 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
-			// Skip processing by xterm.js of keyboard events that resolve to commands described
-			// within commandsToSkipShell
-			if (resolveResult && this._skipTerminalCommands.some(k => k === resolveResult.commandId)) {
+			// for keyboard events that resolve to commands described
+			// within commandsToSkipShell, either alert or skip processing by xterm.js
+			if (resolveResult && this._skipTerminalCommands.some(k => k === resolveResult.commandId) && !this._configHelper.config.sendKeybindingsToShell) {
+				if (this._storageService.getBoolean(SHOW_TERMINAL_CONFIG_PROMPT, StorageScope.GLOBAL, true) &&
+					resolveResult.commandId !== 'workbench.action.terminal.toggleTerminal' &&
+					resolveResult.commandId !== 'workbench.action.terminal.new' &&
+					resolveResult.commandId !== 'workbench.action.togglePanel') {
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:553]: Also `workbench.action.terminal.focus`

**Pull request merged**